lib/sqs.js

```diff
@@ -150,20 +158,162 @@ function createQueue(params) {
+      SentTimestamp: Date.now()
+    },
+    MessageBody, // TODO: assert for all valid message variations
+    MessageId,
+    ...(MessageAttributes && { MessageAttributes }),
+    ...(MessageDeduplicationId && { SequenceNumber: '00000000000000000000' }), // TODO: The length of SequenceNumber is 128 bits, continues to increase for a particular MessageGroupId. Implement seq number,
+    ...(DelaySeconds && { DelaySeconds }),
+    ...(MessageSystemAttributes && { MessageSystemAttributes })
+  };
+
+  const encodedMessage = {
+    MD5OfMessageBody: md5(MessageBody),
+    MessageId,
+    ...(MessageAttributes && { MD5OfMessageAttributes: md5(MessageAttributes) }),
+    ...(MessageDeduplicationId && { SequenceNumber: '00000000000000000000' }),
+    ...(DelaySeconds && { DelaySeconds }),
+    ...(MessageSystemAttributes && { MD5OfMessageSystemAttributes: md5(MessageSystemAttributes) })
+  };
+
+  queue.messages.push(message);
+  return encodedMessage;
+}
+
 function listQueues() {
   const list = [];
   queues.forEach((value, key) => list.push(key));
   return list;
 }
 
+function receiveMessage({
+  AttributeNames,
+  MaxNumberOfMessages = 1,
+  MessageAttributeNames,
+  QueueUrl,
+  VisibilityTimeout
+}) {
+  const convertedQueue = QueueUrl.replace(/http:\/\/localhost:\d{4}\/queues\//g, '');
+  const queue = queues.get(convertedQueue);
+
+  if (!queue) {
+    throw Object.assign(new Error('AWS.SimpleQueueService.NonExistentQueue; see the SQS docs.'), {
+      code: 'AWS.SimpleQueueService.NonExistentQueue'
+    });
+  }
+
+  // if (WaitTimeSeconds) {
+  //   // TODO: amount of time can wait for messages
+  // }
+
+  const stateParams = VisibilityTimeout ? { VisibilityTimeout, isRead: true } : { isRead: true };
+  const messages = queue.messages.slice(0, MaxNumberOfMessages).map(msg => ({
+    ...msg,
+    '@State': {
+      ...msg['@State'],
+      ...stateParams
+    },
+    Attributes: {
+      ...msg.Attributes,
+      ApproximateFirstReceiveTimestamp: Date.now(), // TODO: might need to update only on 1st read
+      ApproximateReceiveCount: msg.Attributes.ApproximateReceiveCount + 1
+    }
+  }));
+
+  const remainingMessages = queue.messages.slice(MaxNumberOfMessages);
+  queue.messages = [...messages, ...remainingMessages];
+
+  return messages.map(({ Attributes, MessageAttributes, MessageBody, MessageId }) => ({
+    Body: MessageBody,
+    MD5OfBody: md5(MessageBody),
+    MessageId,
+    ReceiptHandle: '0000000',
+    ...(AttributeNames && {
+      Attributes: AttributeNames.includes('All')
+        ? Attributes
+        : AttributeNames.reduce((acc, item) => {
+            const value = Attributes[item];
+            if (value === undefined) return acc;
+            return { ...acc, [item]: value };
+          }, {})
+    }),
+    ...(MessageAttributeNames &&
+      MessageAttributeNames.length > 0 && {
+        MD5OfMessageAttributes: md5(MessageAttributeNames.join(',')),
+        MessageAttributes: MessageAttributeNames.includes('All')
```

Comment:
Maybe we should only create an `MD5OfMessageAttributes: md5(...)`
with the result of this line?

Reply:
Anything that is MD5 is more of a placeholder, then the actual working value.